### PR TITLE
parallel link mapping

### PR DESCRIPTION
**Pull request by JacobHayes**
A few attempts to parallelize link mapping across feeds (not routes)

### CODE REVIEW

**Comment by JacobHayes** [comment]
This might be a little easier to follow [without whitespace changes](https://github.com/replicahq/graphhopper/pull/67/files?diff=unified&w=1) (some block indenting w/ the stream/parallel conversions)

**Review by kaelgreco — COMMENTED** [review]
this seems super cool to me! but @mark-idleman  should take a look

do you have any link mapping outputs we can compare? if it's good we can land now.

File: web-bundle/src/main/java/com/graphhopper/replica/GtfsLinkMapper.java
Comment (mark-idleman) [inline web-bundle/src/main/java/com/graphhopper/replica/GtfsLinkMapper.java:126]: take it or leave it nit: IntelliJ suggested `Sets.newHashSet(routeIdToStopPairs.values());` here instead of using `stream`, might be slightly simpler :)

File: web-bundle/src/main/java/com/graphhopper/replica/GtfsLinkMapper.java
Comment (mark-idleman) [inline web-bundle/src/main/java/com/graphhopper/replica/GtfsLinkMapper.java:181]: I'm not sure how big these get - not crazy big, right? Do we think sorting is something that could slow things down enough to matter? If not, sounds awesome!

Comment (mark-idleman) [reply web-bundle/src/main/java/com/graphhopper/replica/GtfsLinkMapper.java:181]: (also, micronit: might be good to add a newline after 181 just to visually denote the end of the big lambda block where we create all the CSV rows)

**Review by mark-idleman — APPROVED** [review]
Epic!!! 🎉  I'm feeling good about this, I know you've done a few tests that looked good recently so if you could share those results here just to get all our eyes on it one more time, that would be awesome. 

Only other request - maybe we could `grpcurl` a single pod running with this to ensure that the new unique key into the `gtfsLinkMappings` map is working perfectly? The code looks great and it's not many changes, just probably good to sanity check given importance. Here's an example request once you're port-forwarded to a pod (replace lat/lons to match area): ``
```
grpcurl -d @ -plaintext localhost:50051 router.Router/RoutePt <<EOM
{
"points":[{"lat":34.10238,"lon":-118.33853},{"lat":34.14835,"lon":-118.2841}],
"earliest_departure_time":"2019-10-13T18:00:00Z",
"limit_solutions":4,
"max_profile_duration":10,
"beta_walk_time":1.5,
"limit_street_time_seconds":1440,
"use_pareto":false,
"betaTransfers":1440000,
}
EOM
```

**Comment by JacobHayes** [comment]
> do you have any link mapping outputs we can compare?

Yes, definitely, thanks for asking! The `sequential` ones (don't have for NA, maybe @mark-idleman has an old one handy) are from the `router-tuning` branch (aka: no changes to this code) and `parallel_v3` are all from the latest commit (https://github.com/replicahq/graphhopper/pull/67/commits/b88c8905b12a0321d9dc90261c802fc1f8c8681e) built with https://github.com/replicahq/model/pull/3841: 
```
> gsutil cp gs://city_data/mini_kc/2020_Q1/graphhopper_graph_builder/9a388276ee25b6fc4fc015dcfcae6ac5352835f0/gtfs_link_mapping.csv mini_kc.sequential.csv
> gsutil cp gs://city_data/mini_kc/2020_Q1/graphhopper_graph_builder/0f18d23269c378e039f25d87432543647fc71d94/gtfs_link_mapping.csv mini_kc.parallel_v3.csv
> gsutil cp gs://city_data/mini_nor_cal/2019_Q4/graphhopper_graph_builder/c7cbb6d9eb78e2221186e4f9599c4dc5246688e4/gtfs_link_mapping.csv mini_nor_cal.sequential.csv
> gsutil cp gs://city_data/mini_nor_cal/2019_Q4/graphhopper_graph_builder/422f14a96bf125b2bbb9b677b1e5e7c5d9230a25/gtfs_link_mapping.csv mini_nor_cal.parallel_v3.csv
> gsutil cp gs://city_data/north_atlantic/2019_Q4/graphhopper_graph_builder/f2078e704b649522b6a54706b46c8728c978a42a/gtfs_link_mapping.csv north_atlantic.parallel_v3.csv
```

and then I've been using something like this to do some comparisons:
```sh
> $CLOUDSDK_PYTHON -m pip install -U crcmod  # Required for downloading "composite" objects
> gsutil cp gs://city_data/mini_nor_cal/2019_Q4/graphhopper_graph_builder/c7cbb6d9eb78e2221186e4f9599c4dc5246688e4/gtfs_link_mapping.csv mini_nor_cal.sequential.csv
> gsutil cp gs://city_data/mini_nor_cal/2019_Q4/graphhopper_graph_builder/422f14a96bf125b2bbb9b677b1e5e7c5d9230a25/gtfs_link_mapping.csv mini_nor_cal.parallel_v3.csv
> ls -hl mini_nor_cal.*
-rw-r--r-- 1 jacobhayes staff 6.4M Jun  7 23:34 mini_nor_cal.parallel_v3.csv
-rw-r--r-- 1 jacobhayes staff 207M Jun  7 23:34 mini_nor_cal.sequential.csv
> wc -l mini_nor_cal.*
    18190 mini_nor_cal.parallel_v3.csv
   609936 mini_nor_cal.sequential.csv
   628126 total
> for file in mini_nor_cal*.csv; do sort "${file}" | uniq > "${file//.csv}.uniq.csv"; done
> ls -hl mini_nor_cal.*.uniq.csv
-rw-r--r-- 1 jacobhayes staff 6.4M Jun  7 23:34 mini_nor_cal.parallel_v3.uniq.csv
-rw-r--r-- 1 jacobhayes staff 6.4M Jun  7 23:34 mini_nor_cal.sequential.uniq.csv
> wc -l mini_nor_cal.*.uniq.csv
   18190 mini_nor_cal.parallel_v3.uniq.csv
   18190 mini_nor_cal.sequential.uniq.csv
   36380 total
> diff -y --suppress-common-lines mini_nor_cal.*.uniq.csv | wc -l
1594
> vimdiff mini_nor_cal.*.uniq.csv
```

After deduping the original version's files, all of the line differences I've seen are in the `street_edges` column. I think the difference is caused by the original code having a bug in the cache key for the stop IDs ([fixed here](https://github.com/replicahq/graphhopper/pull/67/commits/af692dce22675c20d7951b1bc68d04f03aa58052) w/ @mark-idleman's help), which would basically cause it to use the mapped edges from the _first_ feed w/ a given OD ID pair. I haven't confirmed (by running w/ that cache key fix on the otherwise sequential code) though, so there may be something else lurking.

**Comment by JacobHayes** [comment]
> maybe we could grpcurl a single pod running with this to ensure that the new unique key into the gtfsLinkMappings map is working perfectly?

Good call, I'll make this an opportunity to try out the `run-graphhopper data_configs/mini_nor_cal/2019_Q4/season.yaml` (runs locally) added in https://github.com/replicahq/model/pull/3841 too. 😁 How does this request+response look (SF -> Sacramento)?

```sh
> grpcurl -d @ -plaintext localhost:50051 router.Router/RoutePt <<EOM
    {
        "betaTransfers":1440000,
        "beta_walk_time":1.5,
        "earliest_departure_time":"2019-10-13T18:00:00Z",
        "limit_solutions":4,
        "limit_street_time_seconds":1440,
        "max_profile_duration":10,
        "points":[{"lat":37.7775,"lon":-122.416389},{"lat":38.581667,"lon":-121.494444}],
        "use_pareto":false
    }
EOM
```
```json
{
  "paths": [
    {
      "duration_millis": "33110793",
      "distance_meters": 1782.55423381302,
      "transfers": 1,
      "foot_legs": [
        {
          "departure_time": "2019-10-14T00:16:27Z",
          "arrival_time": "2019-10-14T00:30:00Z",
          "distance_meters": 1128.6675956077318,
          "stable_edge_ids": [
            "13453621094869909665",
            "11492053511252861646",
            "17153150084233888469",
            "11799603835779677724",
            "10791231430685387538",
            "11255871766095048535",
            "11611930509958216701",
            "12288521494469474683",
            "14519871810811496591"
          ],
          "travel_segment_type": "ACCESS"
        },
        {
          "departure_time": "2019-10-14T03:04:00Z",
          "arrival_time": "2019-10-14T03:11:50Z",
          "distance_meters": 653.886638205288,
          "stable_edge_ids": [
            "17898505717660857324",
            "10244236554531487638",
            "17449646636992337461",
            "17894474611066647206",
            "12840619790581106693"
          ],
          "travel_segment_type": "EGRESS"
        }
      ],
      "pt_legs": [
        {
          "departure_time": "2019-10-14T00:30:00Z",
          "arrival_time": "2019-10-14T01:15:00Z",
          "trip_id": "37fa3032-17ae-4c1f-a55c-c1e9b79a6d98",
          "route_id": "SF",
          "agency_name": "Capitol Corridor Joint Powers Authority",
          "route_short_name": "SF",
          "route_long_name": "Amtrak Thruway Connecting Service",
          "route_type": "3",
          "direction": "Amtrak Thruway Connecting Service Emeryville",
          "stops": [
            {
              "stop_id": "capitol-corridor:SFS",
              "stop_name": "San Francisco - Shopping Center Amtrak",
              "arrival_time": "1970-01-01T00:00:00Z",
              "departure_time": "2019-10-14T00:30:00Z",
              "point": {
                "lat": 37.78472,
                "lon": -122.406918
              }
            },
            {
              "stop_id": "capitol-corridor:SFC",
              "stop_name": "San Francisco - Transbay Terminal Amtrak",
              "arrival_time": "2019-10-14T00:40:00Z",
              "departure_time": "2019-10-14T00:40:00Z",
              "point": {
                "lat": 37.789232,
                "lon": -122.392366
              }
            },
            {
              "stop_id": "capitol-corridor:EMY",
              "stop_name": "Emeryville",
              "arrival_time": "2019-10-14T01:15:00Z",
              "departure_time": "1970-01-01T00:00:00Z",
              "point": {
                "lat": 37.840423,
                "lon": -122.291808
              }
            }
          ]
        },
        {
          "departure_time": "2019-10-14T01:26:00Z",
          "arrival_time": "2019-10-14T03:04:00Z",
          "trip_id": "742",
          "route_id": "CC",
          "agency_name": "Capitol Corridor Joint Powers Authority",
          "route_short_name": "CC",
          "route_long_name": "Capitol Corridor",
          "route_type": "2",
          "direction": "Capitol Corridor Auburn",
          "stops": [
            {
              "stop_id": "capitol-corridor:EMY",
              "stop_name": "Emeryville",
              "arrival_time": "1970-01-01T00:00:00Z",
              "departure_time": "2019-10-14T01:26:00Z",
              "point": {
                "lat": 37.840423,
                "lon": -122.291808
              }
            },
            {
              "stop_id": "capitol-corridor:BKY",
              "stop_name": "Berkeley",
              "arrival_time": "2019-10-14T01:30:00Z",
              "departure_time": "2019-10-14T01:30:00Z",
              "point": {
                "lat": 37.867439,
                "lon": -122.300908
              }
            },
            {
              "stop_id": "capitol-corridor:RIC",
              "stop_name": "Richmond",
              "arrival_time": "2019-10-14T01:37:00Z",
              "departure_time": "2019-10-14T01:37:00Z",
              "point": {
                "lat": 37.936766,
                "lon": -122.353896
              }
            },
            {
              "stop_id": "capitol-corridor:MTZ",
              "stop_name": "Martinez",
              "arrival_time": "2019-10-14T02:03:00Z",
              "departure_time": "2019-10-14T02:03:00Z",
              "point": {
                "lat": 38.019079,
                "lon": -122.13883
              }
            },
            {
              "stop_id": "capitol-corridor:SUI",
              "stop_name": "Suisun-Fairfield",
              "arrival_time": "2019-10-14T02:22:00Z",
              "departure_time": "2019-10-14T02:22:00Z",
              "point": {
                "lat": 38.243446,
                "lon": -122.041192
              }
            },
            {
              "stop_id": "capitol-corridor:FFV",
              "stop_name": "Fairfield-Vacaville",
              "arrival_time": "2019-10-14T02:29:00Z",
              "departure_time": "2019-10-14T02:29:00Z",
              "point": {
                "lat": 38.2853976,
                "lon": -121.9680691
              }
            },
            {
              "stop_id": "capitol-corridor:DAV",
              "stop_name": "Davis",
              "arrival_time": "2019-10-14T02:49:00Z",
              "departure_time": "2019-10-14T02:49:00Z",
              "point": {
                "lat": 38.543494,
                "lon": -121.737758
              }
            },
            {
              "stop_id": "capitol-corridor:SAC",
              "stop_name": "Sacramento",
              "arrival_time": "2019-10-14T03:04:00Z",
              "departure_time": "1970-01-01T00:00:00Z",
              "point": {
                "lat": 38.584004,
                "lon": -121.500699
              }
            }
          ]
        }
      ]
    }
  ]
}
```

File: web-bundle/src/main/java/com/graphhopper/replica/GtfsLinkMapper.java
Comment (JacobHayes) [reply web-bundle/src/main/java/com/graphhopper/replica/GtfsLinkMapper.java:181]: NA is 189000 rows, so not bad (now that we fixed the ~33x row dup issue haha).

**Comment by JacobHayes** [comment]
Will wait for you all to review the files+request and for https://github.com/replicahq/graphhopper/pull/65 to be merged before merging this one (actually, may not conflict, so might merge first if the rest looks good to you all)!

**Pull request merged**
